protocols/rendezvous/src/behaviour.rs

```diff
@@ -0,0 +1,846 @@
+use crate::codec::{Cookie, ErrorCode, Namespace, NewRegistration, Registration};
+use crate::handler::{InEvent, OutEvent, RendezvousHandler};
+use crate::{handler, MAX_TTL, MIN_TTL};
+use bimap::BiMap;
+use futures::future::BoxFuture;
+use futures::ready;
+use futures::stream::FuturesUnordered;
+use futures::{FutureExt, StreamExt};
+use libp2p_core::connection::ConnectionId;
+use libp2p_core::identity::error::SigningError;
+use libp2p_core::identity::Keypair;
+use libp2p_core::{Multiaddr, PeerId, PeerRecord};
+use libp2p_swarm::{
+    NetworkBehaviour, NetworkBehaviourAction, NotifyHandler, PollParameters, ProtocolsHandler,
+};
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::iter::FromIterator;
+use std::option::Option::None;
+use std::task::{Context, Poll};
+use std::time::Duration;
+use tokio::time::sleep;
+use uuid::Uuid;
+
+pub struct Rendezvous {
```

Comment:
Yep that is an interesting aspect.

I am worried that it will introduce duplicated code because I think that even if a node only acts as a client, they are expected to return `E_UNAVAILABLE` if another node wants to register with us. Hence, we cannot just use things like `upgrade::DeniedUpgrade` to completely disallow inbound substreams.

Consequently, a `RendezvousPointBehaviour` - which would handle inbound substreams - would have to accept every substream regardless and have a config switch to deny them using above error code. To be spec compliant, every node would therefore always include both behaviours anyway, even if one of them is disabled (side note: does this composition even work? Can we have two behaviours that handle the same protocol?)

Thinking about this more, I think I do see your point in that even if both behaviours are always included, splitting the implementation into two roles might make understanding it easier.

We talked about this internally (cc @da-kami @rishflab) but I may have disregarded the idea back then too fast :)
It is quite some work though so we might not jump on it right away :)